feat: CSS Cafe Menu

naomi-lgbt opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->
General Fixes:
- Ensure editable region at end of code-block is followed by a blank
line.
- Remove apparent duplicate step.
- Address spelling issues.
- Add tests!

Commit:
fix: general bugs

- Ensure editable region at end of code-block is followed by a blank
line.
- Remove apparent duplicate step.
- Address spelling issues.

ojeytonwilliams commented:
> Ensure editable region at end of code-block is followed by a blank
line.

I think this is just another Editor bug.  I'd rather track it https://github.com/freeCodeCamp/freeCodeCamp/issues/39348, than use workarounds in the Markdown.  Is that reasonable?

nhcarrigan commented:
Sure - I'll probably revert them after I get the tests written, though, so I'm not fighting the issue while also trying to develop things.

RandellDawson commented:
What makes you think step 55 is a duplicate?

nhcarrigan commented:
> What makes you think step 55 is a duplicate? It could be, but I am not seeing it when I compare the files.

Step 52 states:

![image](https://user-images.githubusercontent.com/63889819/114591440-bbf17100-9c3e-11eb-9c51-62d250c9b50d.png)

Step 54 states:

![image](https://user-images.githubusercontent.com/63889819/114591462-c27fe880-9c3e-11eb-8f5a-f3a9db8077a1.png)

In step 52, the Camper is expected to change:

```css
.flavor {
  text-align: left;
  width: 75%;
}
```

to:

```css
.flavor, .dessert {
  text-align: left;
  width: 75%;
}
```

This change is persisted in Step 53, but then reverted in step 54 and the camper is asked to do it again.

Commit:
feat: some tests

nhcarrigan commented:
Okay, I pushed up some tests for the first few challenges - before I dive in to this full-force, is this the level of granularity and direction we're looking for?

Commit:
feat: tests 6-20

Add tests for steps 6-20.

Commit:
feat: tests 21-30

Add tests for steps 21 through 30.

ojeytonwilliams commented:
> Okay, I pushed up some tests for the first few challenges - before I dive in to this full-force, is this the level of granularity and direction we're looking for?

I like that level of granularity - it goes into a good amount of detail, but there aren't an absurd number of steps.

As for direction, I think we want to talk directly to the user in the description and instructions, but talk about the code in the hints.  I've started a discussion about it in the curriculum chat, but the gist is that that should be a little more encouraging.

Commit:
feat: tests 31-40

Add tests for steps 31 through 40.

Commit:
feat: tests 41-50

Add tests for steps 41 through 50.

Commit:
feat: tests 51-60

Add tests for steps 51 through 60.

Commit:
feat: tests 61-70

Add tests for steps 61 through 70.

Commit:
feat: tests 71-80

Add tests for steps 71 through 80.

Commit:
feat: steps 81-92

Add tests for steps 81 through 92.

Commit:
feat: align dashed names

Align the dashed names with the titles and file names.

RandellDawson commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-011.md:
````diff
@@ -11,10 +11,22 @@ Up until now, you have been limited regarding the presentation and appearance of
+Your `style` element should be nested in your `head` element.
+
+```js
+assert(code.match(/<head>[\w\W\s]*<style>[\w\W\s]*<\/style>[\w\W\s]*<\/head>/i))
````
@ojeytonwilliams At one point you were going to attempt to fix the issue where we can not easily query the head and style elements without having to use regex to find them.

ojeytonwilliams replied:
I did look into that.  As I recall, the problem is that browsers automatically create elements and re-order others, so you can't always know what the user did.  It's been a while, but I think `<title>`s get put into `<head>`s because that's the only permitted parent.  That sort of thing.

We can probably do better than our current approach, but I think there will always be oddities.

gikf reviewed (commented):
Couple of findings and comments.

nhcarrigan commented:
Thanks @gikf for the thorough review - going to give this another pass (I have some more edge case tests to write too).

RandellDawson commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-052.md:
````diff
@@ -7,16 +7,16 @@ dashedName: part-52
 
 ```js
-
+assert(code.match(/\.flavor,\s*.dessert\s*{/i));
````
It would be best to implement a CSS parser instead of writing regex here.

Commit:
fix: apply gikf's suggestions

Co-authored-by: gikf <60067306+gikf@users.noreply.github.com>

nhcarrigan commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-019.md:
````diff
@@ -11,10 +11,23 @@ The text is centered again so the link to the CSS file is working. Add another s
+assert(bodyBackground === 'rgb(165, 42, 42)');
 ```
 
 # --seed--
````
Think this is part of the editable region bug with the editor.

ShaunSHamilton replied:
P.S. I think this is mitigated by leaving a space between the end of the code block on the `--fcc-editable-region--`

ShaunSHamilton replied:
Bump for reminder

nhcarrigan replied:
Fixed in next commit.

Commit:
fix: manual review suggestions

Apply gikf's other suggestions.

nhcarrigan commented:
Going to apply some additional test recommendations from Randy, and sweep through and update the test verbiage at Oliver's recommendation.

Commit:
feat: more thorough testing

When a test asserts that a camper adds a new element relative to
an existing element, also test that the existing element is not
malformed.

moT01 commented:
Are we removing whitespace in the users code for these tests? If so, ignore the rest of this.

One thing I noticed here, that I kind of mentioned in our discussion on the hint verbiage, is that there is a lot of edge cases that the tests won't catch. There's always going to be edge cases, but here's an example:

```
There should be a closing `<title>` tag.
assert(code.match(/<\/title>/i));
```

I believe spaces are valid in tags like this: `</   title   >`. So this test would fail with any spaces in the users code even though it's valid.

Just kind of pointing it out in case it's an issue that should try and be corrected moving forward. Perhaps you are aware or maybe I'm missing some context or something. Ideally, I think we should try and allow any valid code to pass.

ojeytonwilliams commented:
> Ideally, I think we should try and allow any valid code to pass.

I agree.  It makes sense to be strict if we're specifically testing for coding style, but otherwise we don't want to nitpick if we can avoid it.

Commit:
feat: optional semi colons

Make EOL semi-colons optional in the CSS. If they're missing and the
CSS doesn't render correctly, other tests will catch that. If they
are on the last property of the ruleset, they technically aren't
required so we shouldn't require them?

nhcarrigan commented:
Okay I think this is ready for another pass. 

/cc @ahmadabdolsaheb

gikf commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-018.md:
````diff
@@ -17,10 +17,24 @@ Add the following within the `head` element:
+
+```js
+const meta = $('meta');
+assert(meta[0].outerHTML.match(/content=['"]width=device-width, initial-scale=1.0['"]/) || meta[1].outerHTML.match(/content=['"]width=device-width, initial-scale=1.0['"]/));
````
```suggestion
assert(meta[0].outerHTML.match(/content=('|")width=device-width, initial-scale=1.0\1/) || meta[1].outerHTML.match(/content=('|")width=device-width, initial-scale=1.0\1/));
```

gikf commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-018.md:
````diff
@@ -17,10 +17,24 @@ Add the following within the `head` element:
 
+```js
+const meta = $('meta');
+assert(meta[0].outerHTML.match(/name=['"]viewport['"]/) || meta[1].outerHTML.match(/name=['"]viewport['"]/));
````
```suggestion
assert(meta[0].outerHTML.match(/name=('|")viewport\1/) || meta[1].outerHTML.match(/name=('|")viewport\1/));
```

gikf commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-002.md:
````diff
@@ -11,10 +11,34 @@ Add a `head` element within the `html` element, so you can add a `title` element
+assert(code.match(/<\s*title\s*>/i));
+```
+
+You should have a closing `<title>` tag.
````
```suggestion
You should have a closing `</title>` tag.
```

gikf reviewed (commented):
I haven't tried yet completing challenges from start to the end. So far marked are just small things.

I'm not sure if space should be allowed between tag opening character, slash (only in closing tags) and tag name, Preview doesn't like for example `< h1>`, `</ h1>` or `< /h1>` and html specification kind of also doesn't mention space in those places https://html.spec.whatwg.org/multipage/syntax.html#start-tags

nhcarrigan commented:
I'll drop the commit for the spaces between tag names. 🙂

Commit:
chore: apply review suggestions

Apply review suggestions from gikf

Co-authored-by: gikf <60067306+gikf@users.noreply.github.com>

gikf commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-035.md:
````diff
@@ -11,10 +11,22 @@ Next, you want to align the price to the right. Add a class named `price` to you
+You should add the `price` class to your `p` element.
 
 ```js
+assert(code.match(/<p\s*class=['"]price['"]\s*>/i));
````
```suggestion
assert(code.match(/<p\s*class=('|")price\1\s*>/i));
```

gikf commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-037.md:
````diff
@@ -13,10 +13,22 @@ To get them on the same line, you need to apply some styling to the `p` elements
+You should apply the `item` class to your `article` element.
 
 ```js
+assert(code.match(/<article\s*class=['"]item['"]\s*>/i))
````
```suggestion
assert(code.match(/<article\s*class=('|")item\1\s*>/i))
```

gikf commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-062.md:
````diff
@@ -1,20 +1,33 @@
+You should set the `class` of the `p` element to `established`.
 
 ```js
+assert(code.match(/<p class=['"]established['"]>/i));
````
```suggestion
assert(code.match(/<p class=('|")established\1>/i));
```

gikf commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-088.md:
````diff
@@ -1,22 +1,32 @@
+You should apply the `class="address"` attribute.
+
+```js
+assert(code.match(/class=('|")address('|")/i));
````
```suggestion
assert(code.match(/class=('|")address\1/i));
```
This one was suggested incorrectly  :eyes:. Sorry!

gikf reviewed (commented):
Seems like a couple got lost in the process? :slightly_smiling_face:

moT01 reviewed (dismissed):
I would fine merging this. I basically told Randell at one point that I am pretty much fine merging anything to these projects since they won't be live for a while and they will be a work in progress. Things will be pointed out as people test them, and they can be fixed/improved later.

I didn't go through the whole thing, I went through most of it. There's definitely some goofy lessons in there, possibly from some of my previous suggestions, that could be adjusted or reworded. This PR really mostly just adds tests. For the most part, they worked fine. I still have concerns around them not allowing spaces where it's valid in many spots, but I won't hold up the PR for that.

So whenever you think it's good @nhcarrigan - I would say merge it.

nhcarrigan commented:
Thanks @moT01!

I pushed up a draft commit to explore implementing a CSS parser for the tests. If we like it, I can either continue this PR or drop that commit to merge this and work on the parser in a separate PR.

ojeytonwilliams commented on client/src/client/frame-runner.js:
```diff
@@ -1,8 +1,10 @@
+import cssTree from 'css-tree';
 
 window.$ = jQuery;
+window.cssTree = cssTree;
```
If you put this in the same place as curriculumHelpers it'll be available during testing, but you won't be able to use it in a challenge.  You can confirm this by entering `<script> console.log(window.$) </script>` into a challenge and comparing that with `<script> console.log(window.__helpers) </script>`

nhcarrigan replied:
It's 989kb unpacked (not sure if that's big or not).

I could load it in to helpers, sure - my thought process was that being able to access `cssTree` instead of `__helpers.cssTree` might be a smoother DX, but if there's concerns about loading it into the window I can change that. 🙂

ojeytonwilliams reviewed (commented):
Looks good to me.  The only things I'd keep in mind are that we don't want to add a massive library (idk how big css-tree is) if there are smaller options and that we probably want to embed it in the helpers, since it's doing the same job.

ojeytonwilliams commented:
Testing it here makes sense, but I'd prefer adding the parser in a separate PR.  Just in case!

Commit:
chore: apply suggestions from code review

Co-authored-by: gikf <60067306+gikf@users.noreply.github.com>

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-002.md:
````diff
@@ -11,10 +11,34 @@ Add a `head` element within the `html` element, so you can add a `title` element
+```js
+assert(code.match(/<title>/i));
+```
+
````
We should probably check if title is nested in head.

nhcarrigan replied:
Done :)

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-010.md:
````diff
@@ -11,10 +11,36 @@ Create an `h2` element in the `section` element and give it the text `Coffees`.
+assert(h2.parentElement.tagName === 'SECTION');
+```
+
+Your `h2` element should have the text `Coffees`.
````
the tests are perfect. I am not sure if coffees is a word :)

nhcarrigan replied:
Replaced it :D

Commit:
feat: review suggestions

Apply Ahmad's review suggestions.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-014.md:
````diff
@@ -19,10 +19,16 @@ Use a single type selector to center the `h1`, `h2` and `p` elements at the same
+You should use a single type selector for all three elements.
 
 ```js
+assert(code.match(/(h1|h2|p), (h1|h2|p), (h1|h2|p) {/))
````
Should we check for closing brackets?

nhcarrigan replied:
Thoughts on an approach? I'm thinking, based on the order of the tests, I should add `.*}` and allow anything between the braces for this step.

ahmadabdolsaheb replied:
that makes sense.

ahmadabdolsaheb replied:
I see you point. this could also match } of the next selector. I think it is fine as it is

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-015.md:
````diff
@@ -11,10 +11,25 @@ You have styled three elements by writing CSS inside the `style` tags. This work
 ```js
+const css = code.split('<!DOCTYPE html>')[0];
+assert(css.match(/(h1|h2|p), (h1|h2|p), (h1|h2|p) {/))
+```
````
test should also check for closing brackets.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-016.md:
````diff
@@ -11,10 +11,18 @@ Now that you have the CSS in the `styles.css` file, go ahead and remove the `sty
+assert(!code.match(/style/i));
+```
+
+You should not have any CSS selectors in your HTML file.
````
I think this challenge asks for "rewriting the styles you have created into the styles.css file." and the next challenge will ask for removing the css from html.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-016.md:
```diff
@@ -11,10 +11,18 @@ Now that you have the CSS in the `styles.css` file, go ahead and remove the `sty
 # --hints--
 
-Test 1
+You should not have any `style` tags in your code.
```
just a general note: by removing the style element from the from the html, the css styles are still applied and the text is still centered.

nhcarrigan replied:
That's an editor bug, right?

RandellDawson replied:
@ojeytonwilliams @ahmadabdolsaheb The original intent was if the CSS in the `styles.css` tab but the applicable `rel` element was not present in the `index.html` tab and did not link to `styles.css`, then the CSS would not be applied.  This way, it appears and functions as it would in real web development.  Is that not going to happen now?  Having the CSS magically work was something we wanted to avoid.

nhcarrigan replied:
https://github.com/freeCodeCamp/freeCodeCamp/pull/42205 should resolve this.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-017.md:
````diff
@@ -7,14 +7,48 @@ dashedName: part-17
+```js
+assert(code.match(/<head>[\w\W\s]*<link[\w\W\s]*>[\w\W\s]*<\/head>/i))
+```
+
````
tests should check if link is closed properly  />

nhcarrigan replied:
Added a separate test to call that out specifically, and updated this test to include the closing slash.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-020.md:
````diff
@@ -11,10 +11,18 @@ That brown background makes it hard to read the text. Change the `body` element'
 
+```js
+const bodyBackground = $('body').css('background-color');
+console.log(bodyBackground)
````
```suggestion
```

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-026.md:
````diff
@@ -11,10 +11,25 @@ Next, you want to center the `div` horizontally. You can do this by setting its
+```js
+const divLeft = $('div').css('margin-left');
+const divRight = $('div').css('margin-right');
+console.log(parseInt(divLeft))
````
```suggestion
```

nhcarrigan replied:
Whoops! Missed a couple of debugging statements 😅

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-029.md:
````diff
@@ -13,10 +13,23 @@ Delete the comment and its contents inside the `body` type selector. Now add a `
+
+```js
+const background = $('body').css('background-image');
+assert(background === 'url("https://tinyurl.com/coffee-beans-fcc")');
````
general note: we need to self-host this.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-047.md:
```diff
@@ -11,10 +11,28 @@ You will come back to the styling the menu in a few steps, but for now, go ahead
 
```
"desserts" should be replaced by "coffee" in the description?

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-058.md:
```diff
@@ -1,22 +1,28 @@
 # --description--
 
-The current width of the menu will always take of 80% of the `body` element's width. On a very wide screen, the coffee and dessert appear far apart from their prices.
+You can change the `font-family` of text, to make it look different the the default font of your browser. Each browser has some common fonts available to it.
```
```suggestion
You can change the `font-family` of text, to make it look different from the default font of your browser. Each browser has some common fonts available to it.
```

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-067.md:
````diff
@@ -1,20 +1,41 @@
+You should add an `hr` element. `hr` elements are self-closing.
+
+```js
+assert(code.match(/<hr\s?\/?>/i));
````
this would also pass <hr></hr>

nhcarrigan replied:
Added a test to not match `</hr>`.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-067.md:
```diff
@@ -1,20 +1,41 @@
-Add a second `p` element below the one with the link and give it the text `123 freeCodeCamp Drive`.
+You can use an `hr` element to display a divider between sections of different content.
+
+First, add an `hr` element between the first `header` element and the `main` element.
```
if hr elements are not previously taught, we should mention that they are self-closing

nhcarrigan replied:
It's mentioned in the first test, but I included a note in the instructions too. 🙂

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-068.md:
````diff
@@ -1,22 +1,34 @@
+Your `hr` element should have a height of `3px`.
+
+```js
+assert($('hr').css('height') === '3px');
````
`hr {height: 3px;}` does not seem to pass.

nhcarrigan replied:
This passes on my end - what browser are you using?

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-071.md:
````diff
@@ -1,22 +1,28 @@
+Your `hr` element should have a `height` of `2px`.
 
+```js
+assert($('hr').css('height') === '2px');
````
this does not pass with the seed from the next challenge:
```css
hr {
  height: 2px;
  background-color: brown;
  border-color: brown;
}

```

nhcarrigan replied:
Hmm, the height assertions break on zoom levels other than 100% - I see the issue now.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-072.md:
````diff
@@ -1,22 +1,27 @@
+
+```js
+assert($('hr')[1].previousElementSibling.tagName === 'MAIN');
+assert($('hr')[1].nextElementSibling.tagName === 'FOOTER');
````
`<hr></hr>` passes

nhcarrigan replied:
So at this point, we've already taught the camper that `hr` elements are self-closing and we've asserted they use the correct syntax.

I think, rather than asserting it again, it's better to allow code to pass if it renders correctly (and `<hr></hr>` does). Thoughts?

ahmadabdolsaheb replied:
That makes sense.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-083.md:
```diff
@@ -1,22 +1,28 @@
 # --description--
 
-You change properties of a link when the mouse hovers them by using a <dfn>pseudo-selector</dfn> that looks like `a:hover { propertyName: propertyValue; }`.
+You change properties of a link when the link is actually clicked by using a <dfn>pseudo-selector</dfn> that looks like `a:active { propertyName: propertyValue; }`.
```
```suggestion
You change properties of a link when the link is actually being clicked by using a <dfn>pseudo-selector</dfn> that looks like `a:active { propertyName: propertyValue; }`.
```

not sure if this is grammatically more precise. it just made more sense to me.

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-089.md:
```diff
@@ -1,20 +1,34 @@
-Now apply the `address` class to the `p` element containing the address.
+The menu looks good, but other than the coffee beans background image, it is mainly just text.
+
+Under the `Coffees` heading, add an image using the url `https://tinyurl.com/cafe-coffee-fcc`. Give the image an `alt` value of `coffee icon`.
```
Image should be ported to our cdn.

nhcarrigan replied:
https://github.com/freeCodeCamp/cdn/pull/69 🙂

ahmadabdolsaheb commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-092.md:
```diff
@@ -1,20 +1,28 @@
-Add one last image under the `Desserts` heading using the url `https://tinyurl.com/cafe-pie-fcc`. Give the image an `alt` value of `pie icon`.
+It would be nice if the vertical space between the `h2` elements and their associated icons was smaller. The `h2` elements have default top and bottom margin space, so you could change the bottom margin of the `h2` elements to say `0` or another number.
+
+There is an easier way, simply add a negative top margin to the `img` elements to pull them up from their current positions. Negative values are created using a `-` in front of the value. To complete this project, go ahead and use a negative margin of `25px` in the `img` type selector.
```
```suggestion
There is an easier way, simply add a negative top margin to the `img` elements to pull them up from their current positions. Negative values are created using a `-` in front of the value. To complete this project, go ahead and use a negative top margin of `25px` in the `img` type selector.
```
To avoid confusion

ahmadabdolsaheb commented:
@nhcarrigan great job here. I just pointed out some nitpicks. This pr LGMT with or without them.

nhcarrigan commented:
Thanks for the thorough review! I'll get these implemented today 😁

Commit:
chore: apply suggestions from code review

Co-authored-by: Ahmad Abdolsaheb <ahmad.abdolsaheb@gmail.com>

Commit:
chore: additional suggestions

Manual application of Ahmad's review suggestions.

Co-authored-by: Ahmad Abdolsaheb <ahmad.abdolsaheb@gmail.com>

ahmadabdolsaheb commented:
@nhcarrigan feel free to merge this when you are ready.

RandellDawson reviewed (dismissed):
See my comments about making sure the CSS located in the `styles.css` tab is not applied unless the applicable `rel` element links to `styles.css`.

nhcarrigan commented:
Is that an issue with the way the challenge itself is set up? Or is that a bug with the editor?

If it's an editor bug, we're looking to resolve those separately I believe.

RandellDawson commented:
It is more of a platform bug.  It is related to how currently, the styling in the `styles.css` file gets applied regardless if it is "linked".

ojeytonwilliams commented:
> It is more of a platform bug.

I thought my ears were burning.  I should have some time to look at these soon :)

ahmadabdolsaheb commented:
I see Tom has added that to the editor issue, so I think this pr is good to go.

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-052.md:
````diff
@@ -7,16 +7,16 @@ dashedName: part-52
 
 ```js
-
+assert(code.match(/\.(flavor,\s*.dessert)|(dessert,\s*.flavor)\s*{/i));
````
```suggestion
assert(__helpers.removeWhiteSpace(code).match(/\.(flavor,.dessert)|(dessert,.flavor){/i));
```
You might be asking what the difference is, but this allows: `.flavor , dessert{`

moT01 replied:
I am always skeptical of that `removeWhiteSpace` function - wouldn't it allow users to pass with any spaces where they shouldn't be? `.fl  avor, .des sert`. Also, these regex's are missing the espace character in front of the period I think.

ShaunSHamilton replied:
> wouldn't it allow users to pass with any spaces where they shouldn't be? `. fl avor, .des sert`

Honestly, I had never even considered these cases 😨

ShaunSHamilton reviewed (changes requested):
Maybe this has been discussed, but I would have thought it best to include space where the Camper is expected to type?

Will look at more later.

nhcarrigan commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-005.md:
````diff
@@ -11,10 +11,34 @@ The name of the cafe is `Camper Cafe`. Add an `h1` element with the name of the
+assert($('h1')[0].parentElement.tagName === "BODY");
+```
+
+Your `h1` element should have the text `Camper Cafe` in capital letters.
````
![image](https://user-images.githubusercontent.com/63889819/117500611-b6cbcd00-af31-11eb-9e34-4c733281245e.png)

This happens when I pick one :)

nhcarrigan commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-031.md:
````diff
@@ -11,10 +11,30 @@ dashedName: part-31
+
+```js
+const secondP = $('article').children('p')[1];
+assert(secondP.innerText.match(/3.00/i));
````
I'm glad one of us is good at regex, because I'm certainly not 😂

Commit:
chore: apply shaun's review suggestions

Co-authored-by: Shaun Hamilton <51722130+ShaunSHamilton@users.noreply.github.com>

Commit:
feat: add clarification

Add a bit of text clarifying the file switch in the editor.

Commit:
chore: replace URLs

Replace the tinyurls with our self-hosted CDN images.

nhcarrigan commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-015.md:
```diff
@@ -11,10 +11,25 @@ You have styled three elements by writing CSS inside the `style` tags. This work
 # --hints--
 
-Test 1
+Your `styles.css` file should have the `h1, h2, p` type selector.
```
https://github.com/freeCodeCamp/freeCodeCamp/pull/41797/commits/98721c505539a3dfc9d6f89e1c1dac40fb4c1c2c

How's this?

ahmadabdolsaheb commented:
@nhcarrigan, is this waiting on anything else?

ShaunSHamilton commented:
Just to keep you all updated: I do plan on re-reviewing this today.

moT01 commented:
I pointed it out a few times - I do think the lack of allowing spaces in a number of the regex tests is going to be problematic for users. I'm not sure if that should be fixed here, or if we just try and keep moving forward - maybe fix them as issues pop up. I would suggest that we try to account for them with future projects.

One other thought I had is on the test text. Wondering if we are being "hinty" enough. The text tells users what went wrong, but not necessary how to get it right. The system with the RDBMS stuff is quite different, but I pretty much always give users the answer as one of the last hints. So, for instance, here's part 68 on this PR:

```
# --description--
The default properties of an `hr` element will make it appear as a thin light grey line.
You can change the height of the line by specifying a value for the `height` property.
Change the height the `hr` element to be `3px`.

# --hints--
You should use the `hr` selector.
You should set the `height` property to `3px`.
Your `hr` element should have a height of `3px`
```

This one is pretty good, but I'm wondering if we can, or should, somehow give them a hint that maybe says: 
```
In the CSS file, add
hr {
  height: 3px
}
```
or
```
You can add customs styles in your CSS file like this:
<selector> {
  <style-property>: <value>;
}
```
Maybe that's too much.

ShaunSHamilton commented:
> Maybe that's too much

I have been struggling finding the balance. At some point, though, the Camper needs to READ-SEARCH-ASK. If we make it so that a new developer goes through the certification without ever feeling the need to do that, then I think something is wrong.

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-016.md:
```diff
@@ -11,10 +11,18 @@ Now that you have the CSS in the `styles.css` file, go ahead and remove the `sty
 
```
The line above this has a typo `...the text that as centered...`

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-023.md:
````diff
@@ -17,10 +17,18 @@ In your style sheet, comment out the the line containing the `background-color`
+You should comment out the `background-color: burlywood;` line in your CSS.
 
 ```js
+assert(code.match(/\/\*\s*background-color:\s*burlywood;?\s*\*\//i));
````
Do we need to add the flag to not remove comments for this challenge? I forget if that applies to HTML/CSS

nhcarrigan replied:
Might as well put it in there just to be safe.

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-038.md:
```diff
@@ -17,10 +17,22 @@ Using the above selector, add a `display` property with value `inline-block` so
 
```
First sentence has typo: `...p elements are nested in an article elements...`

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-040.md:
```diff
@@ -13,10 +13,16 @@ Change the `width` value to `49%` for each class to see what happens.
 
```
First paragraph typo: `...placing them on a separate lines in...` , `... make each p's width...` - best to change to `...make each p elements width...` for i18n

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-048.md:
```diff
@@ -11,10 +11,22 @@ Add an `h2` element in the new section add give it the text `Desserts`.
 
```
Line 10, typo: `...section add give...`

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-057.md:
```diff
@@ -1,20 +1,28 @@
 # --description--
 
-Since all 4 sides of the menu have the same internal spacing, go ahead and delete the four properties and use a single `padding` property with the value `20px`.
+The current width of the menu will always take of 80% of the `body` element's width. On a very wide screen, the coffee and dessert appear far apart from their prices.
```
```suggestion
The current width of the menu will always take up 80% of the `body` element's width. On a very wide screen, the coffee and dessert appear far apart from their prices.
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-064.md:
```diff
@@ -1,22 +1,39 @@
-The `h1` and `h2` elements' text are set by default values of the user's browser.
-
-Add two new type selectors (`h1` and `h2`). Use the `font-size` property for both, but use the value `40px` for the `h1` and `30px` for the `h2`.
+Add a `footer` element below the `main` element, where you can some additional information.
```
```suggestion
Add a `footer` element below the `main` element, where you can add some additional information.
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-083.md:
```diff
@@ -1,22 +1,28 @@
+You change properties of a link when the link is actually being clicked by using a <dfn>pseudo-selector</dfn> that looks like `a:active { propertyName: propertyValue; }`.
 
-Change the color of the footer `Visit our website` link to be `brown` when a user hovers over it.
+Change the color of the footer `Visit our website` link to be `white` when a clicks on it.
```
```suggestion
Change the color of the footer `Visit our website` link to be `white` when clicked on.
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-083.md:
```diff
@@ -1,22 +1,28 @@
 # --hints--
 
-Test 1
+You should use the `a:active` pseudoselector.
```
```suggestion
You should use the `a:active` pseudo-selector.
```

ShaunSHamilton reviewed (changes requested):
This was my final review for this PR.

I did leave some grammatical errors without comment, because they are insignificant, and beyond the reach of the suggestions.
Also, as Tom said, these tests will be changed from day-1 of prod, if they stay as strict as some of them are. I really do think we need to prioritise a CSS/HTML parser for testing.

Commit:
feat: new CSS tests

Re-write CSS tests with new parser.

Commit:
fix: grammar

Apply Shaun's grammar reviews.

Co-authored-by: Shaun Hamilton <51722130+ShaunSHamilton@users.noreply.github.com>

Commit:
fix: optional chaining

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-003.md:
```diff
@@ -13,10 +13,22 @@ Inside the `head` element, nest a `meta` element with an attribute named `charse
 
```
@nhcarrigan This still needs to be resolved

nhcarrigan replied:
Added to current commit.

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-003.md:
````diff
@@ -13,10 +13,22 @@ Inside the `head` element, nest a `meta` element with an attribute named `charse
+Your `meta` tag should have a `charset` attribute.
+
+```js
+assert(code.match(/<meta charset=/i));
````
There is no check that the `meta` element is within the `head`

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-005.md:
```diff
@@ -11,10 +11,34 @@ The name of the cafe is `Camper Cafe`. Add an `h1` element with the name of the
 
```
@nhcarrigan Also, the instructions are not in-line with the HINTS - capatilisation

nhcarrigan replied:
Both will be fixed in the next commit.

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-009.md:
````diff
@@ -11,10 +11,29 @@ There will be two sections on the menu, one for coffees and one for desserts. Ad
+You should have a closing `</section>` tag.
+
+```js
+assert(code.match(/<\/section>/i));
````
```suggestion
assert(code.match(/<\/section\s*>/i));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-009.md:
````diff
@@ -11,10 +11,29 @@ There will be two sections on the menu, one for coffees and one for desserts. Ad
+You should have an opening `<section>` tag.
 
 ```js
+assert(code.match(/<section>/i));
````
```suggestion
assert(code.match(/<section\s*>/i));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-009.md:
````diff
@@ -11,10 +11,29 @@ There will be two sections on the menu, one for coffees and one for desserts. Ad
+Your `section` element should be within your `main` element.
+
+```js
+const main = document.querySelectorAll('main')[0];
````
```suggestion
const main = document.querySelectorAll('main')?.[0];
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-010.md:
````diff
@@ -7,14 +7,40 @@ dashedName: part-10
+You should have an opening `<h2>` tag.
 
 ```js
+assert(code.match(/<h2>/i));
````
```suggestion
assert(code.match(/<h2\s*>/i));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-010.md:
````diff
@@ -7,14 +7,40 @@ dashedName: part-10
+You should have a closing `</h2>` tag.
+
+```js
+assert(code.match(/<\/h2>/i));
````
```suggestion
assert(code.match(/<\/h2\s*>/i));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-011.md:
````diff
@@ -11,10 +11,22 @@ Up until now, you have been limited regarding the presentation and appearance of
+Your code should have an opening `<style>` tag.
 
 ```js
+assert(code.match(/<style>/i));
````
```suggestion
assert(code.match(/<style\s*>/i));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-011.md:
````diff
@@ -11,10 +11,22 @@ Up until now, you have been limited regarding the presentation and appearance of
+Your code should have a closing `</style>` tag.
 
+```js
+assert(code.match(/<\/style>/));
````
```suggestion
assert(code.match(/<\/style\s*>/));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-011.md:
````diff
@@ -11,10 +11,22 @@ Up until now, you have been limited regarding the presentation and appearance of
+Your `style` element should be nested in your `head` element.
+
+```js
+assert(code.match(/<head>[\w\W\s]*<style>[\w\W\s]*<\/style>[\w\W\s]*<\/head>/i))
````
```suggestion
assert(code.match(/<head\s*>[\w\W\s]*<style\s*>[\w\W\s]*<\/style\s*>[\w\W\s]*<\/head\s*>/i))
```

ShaunSHamilton reviewed (changes requested):
Some things still needing changed, as well as some of the regex for HTML tags not accepting valid syntax (white space)

Commit:
chore: apply suggestions from code review

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

Commit:
chore: apply reviews manually

Commit:
chore: getStyle update

Replace getStyleDeclaration with getStyle.

Commit:
fix: resolve test issue

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-011.md:
````diff
@@ -11,10 +11,22 @@ Up until now, you have been limited regarding the presentation and appearance of
+assert(code.match(/<\/style\s*>/));
+```
+
+Your `style` element should be nested in your `head` element.
````
This is mostly an _out-of-interest_ question: Why have the editable region be more than just the `head` element?

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-017.md:
````diff
@@ -7,14 +7,54 @@ dashedName: part-17
+Your code should have a `link` element.
 
 ```js
+const link = $('link')[1];
+assert(link);
````
```suggestion
// link is removed -> if exists, replaced with style
const link = document.querySelector('body > style');
assert(link);
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-017.md:
````diff
@@ -7,14 +7,54 @@ dashedName: part-17
+Your `link` element should have a `rel` attribute with the value `stylesheet`.
+
+```js
+const link = $('link')[1];
+assert(link.outerHTML.match(/rel=('|")stylesheet\1/i));
````
```suggestion
assert(code.match(/rel=('|")stylesheet\1/i));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-017.md:
````diff
@@ -7,14 +7,54 @@ dashedName: part-17
+Your `link` element should have a `type` attribute with the value `text/css`.
 
+```js
+const link = $('link')[1];
+assert(link.outerHTML.match(/type=('|")text\/css\1/i));
````
```suggestion
assert(code.match(/type=('|")text\/css\1/i));
```

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-017.md:
````diff
@@ -7,14 +7,54 @@ dashedName: part-17
+Your `link` element should have an `href` attribute with the value `styles.css`.
+
+```js
+const link = $('link')[1];
+assert(link.outerHTML.match(/href=('|")styles.css\1/i));
````
```suggestion
assert(code.match(/href=('|")styles.css\1/i));
```

ShaunSHamilton reviewed (changes requested):
What did I tell you about using jQuery in production!

Commit:
chore: remove jQuery :(

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-063.md:
```diff
@@ -1,20 +1,44 @@
 # --description--
 
-Now apply the `established` class to the `Est. 2020` text.
+The `h1` and `h2` elements' text are set by default values of the user's browser.
```
```suggestion
The typography of heading elements (e.g. `h1`, `h2`) is set by default values of users' browsers.
```

OR
```
The heading elements' font sizes are set by default values from the browser.
```

ShaunSHamilton reviewed (dismissed):
This sentence is confusing to me, let alone a new developer. Are we trying to talk about all the default values given to heading elements (typography)? Or, just the font-size?

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-079.md:
```diff
@@ -146,6 +169,8 @@ h1, h2 {
+
 --fcc-editable-region--
 
 --fcc-editable-region--
```
```suggestion
--fcc-editable-region--

```
This helps prevent this:
![image](https://user-images.githubusercontent.com/51722130/121773596-d67b9400-cb74-11eb-93da-0f75c86abb0e.png)

nhcarrigan replied:
This is a feature, not a bug. 😉

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-080.md:
```diff
@@ -1,20 +1,36 @@
 # --description--
 
-Moving down to the `footer` element, make all the text have a value of `14px` for the font size.
+The default color of link that has not yet been clicked on a page is typically blue. The default color of a link that has already been visited from a page is typically purple.
```
```suggestion
The default color of a link that has not yet been clicked on is typically blue. The default color of a link that has already been visited from a page is typically purple.
```

Commit:
chore: apply suggestions from code review

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

Commit:
fiux: Update curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-073.md

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

moT01 reviewed (approved):
LGTM 🎉 I will merge this when the tests pass 👍

ShaunSHamilton commented on curriculum/challenges/english/01-responsive-web-design/basic-css-cafe-menu/part-073.md:
````diff
@@ -1,20 +1,26 @@
+Your `body` element should have a `padding` of `20px`.
 
+```js
+assert($('body').css('padding') === '20px');
````
Yes.... REMOVE THE jQuery!!
